Mathlib/Computability/NFA.lean

```diff
@@ -104,6 +152,150 @@ def accepts : Language α := {x | ∃ S ∈ M.accept, S ∈ M.eval x}
 theorem mem_accepts {x : List α} : x ∈ M.accepts ↔ ∃ S ∈ M.accept, S ∈ M.evalFrom M.start x := by
   rfl
 
+variable (M) in
+/-- `M.Path` represents a concrete path through the NFA from a start state to an end state
+for a particular word.
+
+Note that due to the non-deterministic nature of the automata, the can be more than one `Path`
+for a given word.
+
+Also note that this is `Type` and not a `Prop`, so that we can speak about the properties
+of a particular `Path`, such as the set of states visited along the way (defined as `Path.supp`). -/
+inductive Path : σ → σ → List α → Type (max u v)
+  | nil (s : σ) : Path s s []
+  | cons (t s u : σ) (a : α) (x : List α) :
+      t ∈ M.step s a → Path t u x → Path s u (a :: x)
+
+/-- Set of states visited by a path. -/
+@[simp]
+def Path.supp [DecidableEq σ] {s t : σ} {x : List α} (p : M.Path s t x) : Finset σ :=
+  match p with
+  | nil s => {s}
+  | cons _ _ _ _ _ _ p => {s} ∪ p.supp
+
+theorem Path.evalFrom_iff_nonempty {s t : σ} {x : List α} :
+    t ∈ M.evalFrom {s} x ↔ Nonempty (M.Path s t x) where
+  mp h := match x with
+    | [] =>
+      have h : s = t := by simp at h; tauto
+      ⟨h ▸ Path.nil s⟩
+    | a :: x =>
+      have h : ∃ s' ∈ M.step s a, t ∈ M.evalFrom {s'} x :=
+        by rw [evalFrom_cons, mem_evalFrom_iff_exists, stepSet_singleton] at h; exact h
+      let ⟨s', h₁, h₂⟩ := h
+      let ⟨p'⟩ := Path.evalFrom_iff_nonempty.1 h₂
+      ⟨Path.cons s' _ _ _ _ h₁ p'⟩
+  mpr p := match p with
+    | ⟨nil s⟩ => by simp
+    | ⟨cons s' s t a x h₁ h₂⟩ => by
+      rw [evalFrom_cons, stepSet_singleton, mem_evalFrom_iff_exists]
+      exact ⟨s', h₁, Path.evalFrom_iff_nonempty.2 ⟨h₂⟩⟩
+
+theorem Path.accepts_iff_nonempty {x : List α} :
```

Comment:
```suggestion
theorem accepts_iff_exists_path {x : List α} :
```
follows the naming convention and is suggestive

Reply:
Okay, I assume I should also rename `Path.evalFrom_iff_nonempty` to `mem_evalFrom_iff_exists_path`.